[SimHanKiong] iP

SimHanKiong opened the pull request:
### Johnny

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Johnny helps you manage your tasks. It is

- Text-based
- Easy to learn
- ~~Fast~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/SimHanKiong/ip/releases/tag/A-Jar).
2. double-click it
3. add your tasks
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon) 

Here's the `main` method:

``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

SimHanKiong committed:
Reached Level 0

SimHanKiong committed:
Reached Level 1

SimHanKiong committed:
Reached Level 2

SimHanKiong committed:
Reached Level 3

SimHanKiong committed:
Reach Level 4

SimHanKiong committed:
Reached A-TextUiTesting

SimHanKiong committed:
Reached Level 5

SimHanKiong committed:
Reached Level-6

SimHanKiong committed:
Reached A-Enums

SimHanKiong committed:
Add Save Feature

SimHanKiong committed:
Add data/johnny.txt Storage File

SimHanKiong committed:
Add Date and Time parsing

SimHanKiong committed:
Remove tracking of storage file

SimHanKiong committed:
Merge branch 'branch-Level-8' into 'master'

SimHanKiong committed:
Refactor to match OOP standards

SimHanKiong committed:
Merge branch 'branch-A-MoreOOP'

SimHanKiong committed:
Refactor code into separate packages

SimHanKiong committed:
Merge branch 'branch-A-Packages'

SimHanKiong committed:
Fix package refactor error

SimHanKiong committed:
Merge branch 'branch-A-Packages'

SimHanKiong committed:
Merge commit 'add-gradle-support' into branch-A-Gradle

SimHanKiong committed:
Add Gradle and Refactor code to fit checkstyle requirements

SimHanKiong committed:
Merge branch 'branch-A-Gradle'

SimHanKiong committed:
Add JUnit tests for tasks package

SimHanKiong committed:
Merge branch 'branch-A-JUnit'

SimHanKiong committed:
Refactor to allow jar file to be created

SimHanKiong committed:
Merge branch 'branch-A-Jar'

SimHanKiong committed:
Add JavaDocs to all classes, fields and methods

SimHanKiong committed:
Add JavaDocs to Ui class missing previously

SimHanKiong committed:
Refactor imports in test/task package to comply with checkstyle

SimHanKiong committed:
Add Find command function

SimHanKiong committed:
Merge branch 'branch-Level-9'

SimHanKiong committed:
Merge branch 'branch-A-CodingStandard'

SimHanKiong committed:
Merge branch 'branch-A-JavaDoc'

SimHanKiong committed:
Fix Checkstyle

SimHanKiong committed:
Ran CheckStyle and checked for naming convention errors.

SimHanKiong committed:
Merge branch 'branch-A-CheckStyle'

SimHanKiong committed:
Add GUI

SimHanKiong committed:
Merge branch 'branch-Level-10'

TarinPairor commented on src/main/java/johnny/commands/FindCommand.java:
```diff
@@ -0,0 +1,52 @@
+import johnny.tasks.TaskList;
+import johnny.ui.Ui;
+
+import java.util.jar.JarException;
```
Is this import not used? Perhaps it should be discarded.

SimHanKiong replied:
whoops probably forgot to remove, thanks

TarinPairor commented on src/test/java/johnny/tasks/DeadlineTest.java:
```diff
@@ -1,10 +1,10 @@
 package johnny.tasks;
 
-import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Maybe better to group similar imports together?

```
import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.Test;
```

SimHanKiong replied:
yep, thanks

TarinPairor commented on src/main/java/johnny/ui/Ui.java:
```diff
@@ -54,7 +54,6 @@ public void showList(TaskList tasks) throws JohnnyException {
             }
             System.out.println();
         }
-
```
Was this commit necessary?

TarinPairor replied:
Ok I see the point now.

TarinPairor commented on src/main/java/johnny/commands/FindCommand.java:
```diff
@@ -0,0 +1,52 @@
+     * @return False so the loop keeps running.
+     */
+    @Override
+    public boolean isExit() {
```
Smart to have function but is this necessary? Also if it's a function should the naming convention be something else? A variable isExit makes sense but maybe `shouldExit()` could be better to not be confused for a variable name.

TarinPairor commented on src/main/java/johnny/tasks/TaskList.java:
```diff
@@ -54,4 +54,46 @@ public void addTask(Task task) {
     }
 
+    /**
+     * Find all Tasks in TaskList that name contains keyword.
```
This could be rephrased for clarity. Something like "Finds all tasks in the TaskList whose names contain the specified keyword" might be clearer.

SimHanKiong replied:
Thanks, a lot clearer now

TarinPairor commented on src/main/java/johnny/tasks/TaskList.java:
```diff
@@ -54,4 +54,46 @@ public void addTask(Task task) {
+                return false;
+            }
+            for (int i = 0; i < this.size(); i++) {
+                if (this.tasks.get(i) != t.tasks.get(i)) {
```
```suggestion
      (!this.tasks.get(i).equals(t.tasks.get(i)))
```

TarinPairor commented on src/main/java/johnny/ui/Ui.java:
```diff
@@ -94,18 +92,9 @@ public void showDelete(Task task, TaskList tasks) {
      * @param tasks Tasks left in TaskList.
```
Were there incomplete parameters?

SimHanKiong replied:
caught it, thanks

TarinPairor commented on src/main/java/johnny/Johnny.java:
```diff
@@ -1,5 +1,19 @@
 package johnny;
 
+import javafx.application.Application;
```
Overall very well written and well designed. Hard to find substantial errors.

SimHanKiong replied:
Thanks for being so detailed

SimHanKiong committed:
Refactor Parser class

The previous implementation was one long method. It has now been broken
down into smaller methods for better abstraction.

Error handling is specific to each small method so specific exceptions
are caught at each method.

This helps to improve code quality.

SimHanKiong committed:
Add Assertions to code

Ensures names in tasks are not empty and that output string should
not be empty. This prevents unexpected behaviour in code and makes
the app more robust.

SimHanKiong committed:
Merge pull request #1 from SimHanKiong/branch-A-CodeQuality

Refactor Parser class

SimHanKiong committed:
Merge branch 'master' into branch-A-Assertions

SimHanKiong committed:
Merge pull request #2 from SimHanKiong/branch-A-Assertions

Branch a assertions

SimHanKiong committed:
Merge branch 'master' of https://github.com/SimHanKiong/ip

SimHanKiong committed:
Add CI to project

Adds CI to the project such that every time files a pull request or a
push occurs, the workflow will run automatically.

Runs ./gradlew check every time the event above occurs.

SimHanKiong committed:
Merge branch 'branch-A-CI'

SimHanKiong committed:
Extend mark, unmark and delete to accept multiple parameters

Mark, unmark and delete can now run for multiple indices and will not
throw too many arguments error anymore. This will make marking,
unmarking and deleting faster as multiple tasks can be marked,
unmarked and deleted at once.

SimHanKiong committed:
Merge branch 'branch-C-MassOps'

SimHanKiong committed:
Fix Test for find command

SimHanKiong committed:
Fix Javadoc comments

SimHanKiong committed:
Add User Guide and Screenshot of Ui

SimHanKiong committed:
Fix Gradle Build File
